### PR TITLE
[MRG] Reraise when the exception is not a TransportableException

### PR DESCRIPTION
**Pull request by lesteve**
Usual caveat about catching an exception, raising the exception and losing some information in the stacktrace.

An example I bumped into when writing a custom parallel backend:

```py
from joblib import Parallel, delayed
from joblib.parallel import ParallelBackendBase, parallel_backend


class Result(object):
    def __init__(self, result):
        self.result = result

    def get(self, timeout=None):
        1/0


class MyBackend(ParallelBackendBase):
    def apply_async(self, batch, callback):
        return Result(batch())

    def effective_n_jobs(self, n_jobs):
        return n_jobs


def func(i):
    return i

with parallel_backend(MyBackend(), n_jobs=2):
    Parallel(n_jobs=2)(delayed(func)(i) for i in range(10))
```

On Python 2.7 (on Python 3 the problem is less severe):
master:
```
Traceback (most recent call last):
  File "/tmp/test.py", line 25, in <module>
    Parallel(n_jobs=2)(delayed(func)(i) for i in range(10))
  File "/home/lesteve/dev/joblib/joblib/parallel.py", line 768, in __call__
    self.retrieve()
  File "/home/lesteve/dev/joblib/joblib/parallel.py", line 719, in retrieve
    raise exception
ZeroDivisionError: integer division or modulo by zero
```
this PR:
```
Traceback (most recent call last):
  File "/tmp/test.py", line 25, in <module>
    Parallel(n_jobs=2)(delayed(func)(i) for i in range(10))
  File "/home/lesteve/dev/joblib/joblib/parallel.py", line 771, in __call__
    self.retrieve()
  File "/home/lesteve/dev/joblib/joblib/parallel.py", line 682, in retrieve
    self._output.extend(job.get(timeout=self.timeout))
  File "/tmp/test.py", line 10, in get
    1/0
ZeroDivisionError: integer division or modulo by zero
```

On master you lose the line where the error actually happened. In my case it would have saved me a few hours of being really really confused.

### CODE REVIEW

File: joblib/parallel.py [hunk 1]
```diff
@@ -689,7 +689,22 @@ def retrieve(self):
                 # Stop dispatching any new job in the async callback thread
                 self._aborting = True
 
-                if isinstance(exception, TransportableException):
+                # If the backends allows it, cancel or kill remaining running
```
Comment (aabadie) [inline joblib/parallel.py]: backend

Comment (lesteve) [inline joblib/parallel.py]: Not my typo ;-) (I just moved the code around) but I'll fix it anyway.

Comment (lesteve) [inline joblib/parallel.py]: Done

**Comment by aabadie** [comment]
Thanks @lesteve. Feel free to merge.

**Comment by lesteve** [comment]
> Thanks @lesteve. Feel free to merge.

Can you change your review to "Approved"?

**Comment by aabadie** [comment]
It's done already ;)

**Comment by aabadie** [comment]
Ah, I think you can't merge your own PR. So merginf, thanks !

**Pull request merged**